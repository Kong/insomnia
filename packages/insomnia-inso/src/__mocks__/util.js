// eslint-disable-next-line filenames/match-exported
const mod = jest.requireActual('../util');

mod.exit = jest.fn();

module.exports = mod;
