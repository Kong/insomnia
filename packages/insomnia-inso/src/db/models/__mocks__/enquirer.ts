const __constructorMock = jest.fn();

let returnValue = 'test';

module.exports = {
  __constructorMock,
  __mockPromptRun: (value: string) => {
    returnValue = value;
  },
  AutoComplete: class {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- missing types from enquirer
    constructor(options: any) {
      __constructorMock(options);
    }

    run() {
      return returnValue;
    }
  },
};
