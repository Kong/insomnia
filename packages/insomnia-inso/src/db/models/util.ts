import { InsoError } from '../../errors';
import type { BaseModel } from './types';

export const matchIdIsh = ({ _id }: BaseModel, identifier: string) =>
  _id.startsWith(identifier);

export const generateIdIsh = ({ _id }: BaseModel, length = 10) =>
  _id.substr(0, length);

function indent(level: number, code: string, tab = '  |'): string {
  if (!level || level < 0) {
    return code;
  }

  const prefix = new Array(level + 1).join(tab);
  return `${prefix} ${code}`;
}

export const getDbChoice = (
  idIsh: string,
  message: string,
  config: {
        indent?: number
        hint?: string
    } = {},
) => ({
  name: idIsh,
  message: indent(config.indent || 0, message),
  value: `${message} - ${idIsh}`,
  hint: config.hint || `${idIsh}`,
});

export const ensureSingleOrNone = <T>(
  items: T[],
  entity: string,
): T | null | undefined => {
  if (items.length === 1) {
    return items[0];
  }

  if (items.length === 0) {
    return null;
  }

  throw new InsoError(
    `Expected single or no ${entity} in the data store, but found multiple (${items.length}).`,
  );
};
export const ensureSingle = <T>(items: T[], entity: string): T => {
  if (items.length === 1) {
    return items[0];
  }

  if (items.length === 0) {
    throw new InsoError(
      `Expected single ${entity} in the data store, but found none.`,
    );
  }

  throw new InsoError(
    `Expected single ${entity} in the data store, but found multiple (${items.length}).`,
  );
};
