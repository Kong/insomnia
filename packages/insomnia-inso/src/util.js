// @flow
import * as packageJson from '../package.json';
import { handleError } from './errors';

export function getVersion() {
  return isDevelopment() ? 'dev' : packageJson.version;
}

export function isDevelopment() {
  return process.env.NODE_ENV === 'development';
}

export function logErrorExit1(err?: Error) {
  handleError(err);
  process.exit(1);
}

export async function exit(result: Promise<boolean>): Promise<void> {
  return result.then(r => process.exit(r ? 0 : 1)).catch(logErrorExit1);
}

export function getDefaultAppDataDir(): string {
  const dir = process.env.DEFAULT_APP_DATA_DIR;

  if (!dir) {
    throw new Error('Environment variable DEFAULT_APP_DATA_DIR is not set.');
  }

  return dir;
}
