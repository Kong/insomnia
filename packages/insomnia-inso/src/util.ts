import packageJson from '../package.json';
import { handleError } from './errors';

export const getVersion = () => {
  return isDevelopment() ? 'dev' : packageJson.version;
};

export const isDevelopment = () => {
  return process.env.NODE_ENV === 'development';
};

export const logErrorExit1 = (err?: Error) => {
  handleError(err);
  process.exit(1);
};

export const exit = async (result: Promise<boolean>): Promise<void> => {
  return result.then(r => process.exit(r ? 0 : 1)).catch(logErrorExit1);
};

export const getDefaultAppName = (): string => {
  const name = process.env.DEFAULT_APP_NAME;

  if (!name) {
    throw new Error('Environment variable DEFAULT_APP_NAME is not set.');
  }

  return name;
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const noop = () => {};
