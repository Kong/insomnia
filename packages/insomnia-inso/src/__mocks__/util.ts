import { jest } from '@jest/globals';

import * as utilOriginal from '../util';

// eslint-disable-next-line filenames/match-exported
const util = jest.requireActual('../util') as typeof utilOriginal;
util.exit = jest.fn();

module.exports = util;
