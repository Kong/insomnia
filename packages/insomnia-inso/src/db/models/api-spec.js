// @flow
import type { Database } from '../index';
import type { ApiSpec } from './types';
import { mustFindSingleOrNone } from '../index';
import { generateIdIsh, getDbChoice, matchIdIsh } from './util';
import { AutoComplete } from 'enquirer';

export const loadApiSpec = (db: Database, identifier: string): ?ApiSpec =>
  mustFindSingleOrNone(db.ApiSpec, s => matchIdIsh(s, identifier) || s.fileName === identifier);

export const promptApiSpec = async (db: Database, ci: boolean): Promise<?ApiSpec> => {
  if (ci || !db.ApiSpec.length) {
    return null;
  }

  const prompt = new AutoComplete({
    name: 'apiSpec',
    message: 'Select an API Specification',
    choices: db.ApiSpec.map(s => getDbChoice(generateIdIsh(s), s.fileName)),
  });

  const [idIsh] = (await prompt.run()).split(' - ').reverse();
  return loadApiSpec(db, idIsh);
};
