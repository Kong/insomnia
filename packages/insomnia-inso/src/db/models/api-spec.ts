// @ts-expect-error the enquirer types are incomplete https://github.com/enquirer/enquirer/pull/307
import { AutoComplete } from 'enquirer';

import { logger } from '../../logger';
import type { Database } from '../index';
import type { ApiSpec } from './types';
import {
  ensureSingleOrNone,
  generateIdIsh,
  getDbChoice,
  matchIdIsh,
} from './util';
const entity = 'api specification';

export const loadApiSpec = (
  db: Database,
  identifier: string,
): ApiSpec | null | undefined => {
  logger.trace(
    'Load %s with identifier `%s` from data store',
    entity,
    identifier,
  );
  const items = db.ApiSpec.filter(
    spec => matchIdIsh(spec, identifier) || spec.fileName === identifier,
  );
  logger.trace('Found %d.', items.length);
  return ensureSingleOrNone(items, entity);
};

export const promptApiSpec = async (
  db: Database,
  ci: boolean,
): Promise<ApiSpec | null | undefined> => {
  if (ci || !db.ApiSpec.length) {
    return null;
  }

  const prompt = new AutoComplete({
    name: 'apiSpec',
    message: 'Select an API Specification',
    choices: db.ApiSpec.map(s => getDbChoice(generateIdIsh(s), s.fileName)),
  });
  logger.trace('Prompt for %s', entity);
  const [idIsh] = (await prompt.run()).split(' - ').reverse();
  return loadApiSpec(db, idIsh);
};
