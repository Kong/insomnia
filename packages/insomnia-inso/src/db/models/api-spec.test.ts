import enquirer from 'enquirer';

import type { Database } from '../index';
import { emptyDb } from '../index';
import { loadApiSpec, promptApiSpec } from './api-spec';
import type { ApiSpec } from './types';
import { generateIdIsh } from './util';

jest.mock('enquirer');

describe('apiSpec', () => {
  let db: Database = emptyDb();

  const spec: Partial<ApiSpec> = {
    _id: 'spc_1234567890',
    fileName: 'fileName',
  } as ApiSpec;

  beforeEach(() => {
    db = emptyDb();
    const dummySpec: Partial<ApiSpec> = {
      _id: 'spc_dummy',
      fileName: 'dummy spec',
    } as ApiSpec;
    db.ApiSpec.push(spec);
    db.ApiSpec.push(dummySpec);
    jest.clearAllMocks();
  });

  describe('promptApiSpec()', () => {
    it('should return null if ci', () => {
      expect(promptApiSpec(db, true)).resolves.toBeNull();
    });

    it('should return null if db.ApiSpec is empty', () => {
      db.ApiSpec = [];
      expect(promptApiSpec(db, false)).resolves.toBeNull();
    });

    it('should return null if prompt result does not contain closing [ - id]', async () => {
      enquirer.__mockPromptRun('malformed result');

      expect(promptApiSpec(db, false)).resolves.toBeNull();
    });

    it('should load apiSpec after prompt result', async () => {
      enquirer.__mockPromptRun('fileName - spc_123456');

      const result = await promptApiSpec(db, false);
      expect(result).toBe(spec);
    });

    it('should match snapshot of autocomplete config', async () => {
      await promptApiSpec(db, false);
      expect(enquirer.__constructorMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('loadApiSpec()', () => {
    it('should return null if spec not found', () => {
      expect(loadApiSpec(db, 'not-found')).toBeNull();
    });

    it.each([generateIdIsh(spec), spec._id, spec.fileName])(
      'should return spec with identifier: %o',
      (identifier) => {
        expect(loadApiSpec(db, identifier)).toBe(spec);
      },
    );

    it('should throw error if multiple specs matched', () => {
      db.ApiSpec.push({ ...spec });
      expect(() => loadApiSpec(db, spec._id)).toThrowError();
    });
  });
});
