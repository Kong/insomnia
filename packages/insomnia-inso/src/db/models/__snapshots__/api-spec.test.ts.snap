// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiSpec promptApiSpec() should match snapshot of autocomplete config 1`] = `
Object {
  "choices": Array [
    Object {
      "hint": "spc_123456",
      "message": "fileName",
      "name": "spc_123456",
      "value": "fileName - spc_123456",
    },
    Object {
      "hint": "spc_dummy",
      "message": "dummy spec",
      "name": "spc_dummy",
      "value": "dummy spec - spc_dummy",
    },
  ],
  "message": "Select an API Specification",
  "name": "apiSpec",
}
`;
